### PR TITLE
Csi translation portworx

### PR DESCRIPTION
**Pull request by trierra**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add support for Portworx plugin to csi-translation-lib

in support of csi-migration proposal here: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md

Will help with migration of in-tree Portworx plugin to Portworx CSI.

#### Which issue(s) this PR fixes:

Fixes #
[2589](https://github.com/kubernetes/enhancements/issues/2589)
#### Does this PR introduce a user-facing change?

```release-note

Add support for Portworx plugin to csi-translation-lib. Alpha release

Portworx CSI driver is required to enable migration.
This PR adds support of the `CSIMigrationPortworx` feature gate, which can be enabled by:

1. Adding the feature flag to the kube-controller-manager `--feature-gates=CSIMigrationPortworx=true` 
2. Adding the feature flag to the kubelet config:

featureGates:
  CSIMigrationPortworx: true
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

/sig storage
/priority important-soon

### CODE REVIEW

**Review by Jiawei0227 — COMMENTED** [review]
- cmd/kubelet/app/plugins_providers.go
- cmd/kubelet/app/plugins.go
- cmd/kube-controller-manager/app/plugins.go 
- cmd/kube-controller-manager/app/plugins_providers.go 

These files also need to change. Basically we want to make sure when `InTreePluginPortworxUnregister` is enabled. The intree portworx plugin is unregistered.

**Review by Jiawei0227 — COMMENTED** [review]
I think the other part looks good to me.
/cc @msau42 @jsafrane

**Comment by SergeyKanzhelev** [comment]
/kind feature

it doesn't look like enhancement issue https://github.com/kubernetes/enhancements/issues/2589 is in 1.22 milestone. So likely this is not for tomorrow's code freeze. Perhaps backlog is the right priority in this case:

/priority backlog

sig storage please clear up triage label

File: staging/src/k8s.io/csi-translation-lib/go.mod
Comment (msau42) [inline staging/src/k8s.io/csi-translation-lib/go.mod:8]: It doesn't look like your change modifie/added any dependencies. Is this change necessary?

Comment (trierra) [reply staging/src/k8s.io/csi-translation-lib/go.mod:8]: I don't think so. The `make verify` has failed and I ran `hack/update-vendor.sh` as was suggested in the error message, I think it just pulled some other dependencies. Would love to get an advice how to fix that

Comment (Jiawei0227) [reply staging/src/k8s.io/csi-translation-lib/go.mod:8]: Can you try just delete these if these are not used?

**Review by Jiawei0227 — COMMENTED** [review]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/103447/pull-kubernetes-verify/1412457277926936576

According to this, seems you are having some dependency issue? But since you dont import new codes, maybe just delete these changes? And also, you have 

```
Boilerplate header is wrong for: /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/csi-translation-lib/plugins/portworx.go
```
This error so maybe check the header?

File: staging/src/k8s.io/csi-translation-lib/go.mod
Comment (trierra) [reply staging/src/k8s.io/csi-translation-lib/go.mod:8]: sure, thanks!

**Review by Jiawei0227 — APPROVED** [review]
/lgtm

File: pkg/volume/csimigration/plugin_manager.go
Comment (Huang-Wei) [inline pkg/volume/csimigration/plugin_manager.go:22]: please leave the empty line as-is. (applies elsewhere)

File: pkg/scheduler/framework/plugins/nodevolumelimits/utils.go
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/nodevolumelimits/utils.go:51]: Scheduler changes LGTM.

Will approve once sig-storage approves the feature gate's addition.

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:369]: likely needs to be changed to 1.23 as we are at the code freeze

Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:369]:
```suggestion
	// alpha: v1.23
```

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:375]:
```suggestion
	// alpha: v1.23
```

**Comment by SergeyKanzhelev** [comment]
/lgtm cancel

(cancelling to update the milestones - easier to triage without label)

File: pkg/volume/csimigration/plugin_manager.go
Comment (trierra) [reply pkg/volume/csimigration/plugin_manager.go:22]: I don't see this empty line in code anymore but it still shows up here...

Comment (SergeyKanzhelev) [reply pkg/volume/csimigration/plugin_manager.go:22]: @Huang-Wei asked to return it back so there is an empty line to keep separation of import blocks - stdlib vs. other

Comment (trierra) [reply pkg/volume/csimigration/plugin_manager.go:22]: oh, sorry, I didn't read it right the first time. Fixed

**Comment by msau42** [comment]
@trierra are you able to share results from the storage e2e tests to demonstrate that the basics are working?

**Comment by Jiawei0227** [comment]
gentle ping @trierra , any tests?

**Comment by trierra** [comment]
yes, didn't see the previous comment, thanks for the reminder! Will post the results shortly

**Comment by Jiawei0227** [comment]
/lgtm
If you have already tested this e2e and it is working as expected we will be able to ask for approval on this

**Comment by trierra** [comment]
yes, I have, but it looks like I need to post results here. I was planning to complete that this week. Will that work?

**Comment by Jiawei0227** [comment]
Any updates on this?

**Comment by trierra** [comment]
yes. All e2e tests against the portworx in-tree driver with migration enabled have passed, except PVC expand. Need to update the external-resizer vendor libraries and run tests again.

**Comment by Jiawei0227** [comment]
sg! Can you rebase and we can do a final review on it

**Comment by Jiawei0227** [comment]
Overall looks good to me. Thanks for the effort!

Can you update the release-note to include
1. User need to install portworx csi driver to enable the feature
2. Add description to the new feature gates that get added

/lgtm
/cc @msau42 @gnufied 
for approval

**Comment by trierra** [comment]
logs from external provisioner: shows new volumes, created using the storageclass with in-tree driver, get provisioned successfully by csi-provisioner when CSI migration is enabled 
```
I1109 22:04:39.095760       1 controller.go:858] successfully created PV pvc-82cfbf5e-322c-4785-b421-98324f8e0806 for PVC pgbench-state1-migrated and csi volume name 797625906687706306
I1109 22:04:39.095843       1 controller.go:1384] provision "default/pgbench-state1-migrated" class "postgres-pgbench-sc": volume "pvc-82cfbf5e-322c-4785-b421-98324f8e0806" provisioned
I1109 22:04:39.095866       1 controller.go:1397] provision "default/pgbench-state1-migrated" class "postgres-pgbench-sc": succeeded
I1109 22:04:39.101767       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pgbench-state1-migrated", UID:"82cfbf5e-322c-4785-b421-98324f8e0806", APIVersion:"v1", ResourceVersion:"23888", FieldPath:""}): type: 'Normal' reason: 'ProvisioningSucceeded' Successfully provisioned volume pvc-82cfbf5e-322c-4785-b421-98324f8e0806
I1109 22:04:39.857186       1 controller.go:1279] provision "default/pgbench-data-migrated" class "postgres-pgbench-sc": started
I1109 22:04:39.857262       1 controller.go:520] translating storage class for in-tree plugin kubernetes.io/portworx-volume to CSI
I1109 22:04:39.857611       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pgbench-data-migrated", UID:"a2673fb4-e637-4aac-86b8-af60cad7a21f", APIVersion:"v1", ResourceVersion:"23889", FieldPath:""}): type: 'Normal' reason: 'Provisioning' External provisioner is provisioning volume for claim "default/pgbench-data-migrated"
I1109 22:04:40.301368       1 controller.go:774] create volume rep: {CapacityBytes:10737418240 VolumeId:793797135136654159 VolumeContext:map[attached:ATTACH_STATE_INTERNAL_SWITCH error: parent: readonly:false secure:false shared:false sharedv4:false state:VOLUME_STATE_DETACHED] ContentSource:<nil> AccessibleTopology:[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I1109 22:04:40.301434       1 controller.go:858] successfully created PV pvc-a2673fb4-e637-4aac-86b8-af60cad7a21f for PVC pgbench-data-migrated and csi volume name 793797135136654159
I1109 22:04:40.301457       1 controller.go:1384] provision "default/pgbench-data-migrated" class "postgres-pgbench-sc": volume "pvc-a2673fb4-e637-4aac-86b8-af60cad7a21f" provisioned
I1109 22:04:40.301496       1 controller.go:1397] provision "default/pgbench-data-migrated" class "postgres-pgbench-sc": succeeded
I1109 22:04:40.306991       1 event.go:282] Event(v1.ObjectReference{Kind:"PersistentVolumeClaim", Namespace:"default", Name:"pgbench-data-migrated", UID:"a2673fb4-e637-4aac-86b8-af60cad7a21f", APIVersion:"v1", ResourceVersion:"23932", FieldPath:""}): type: 'Normal' reason: 'ProvisioningSucceeded' Successfully provisioned volume pvc-a2673fb4-e637-4aac-86b8-af60cad7a21f
^C^J
oksana@dev-onaumov ~/g/kubeup (master) [SIGINT]> k get storageclass postgres-pgbench-sc -oyaml
allowVolumeExpansion: true
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"allowVolumeExpansion":true,"apiVersion":"storage.k8s.io/v1","kind":"StorageClass","metadata":{"annotations":{},"name":"postgres-pgbench-sc"},"parameters":{"repl":"1"},"provisioner":"kubernetes.io/portworx-volume"}
  creationTimestamp: "2021-11-09T21:53:54Z"
  name: postgres-pgbench-sc
  resourceVersion: "22531"
  uid: 80617ace-aff4-4755-be00-2f578f7973f1
parameters:
  repl: "1"
provisioner: kubernetes.io/portworx-volume
reclaimPolicy: Delete
volumeBindingMode: Immediate
oksana@dev-onaumov ~/g/kubeup (master)> k get pvc
NAME                      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS          AGE
pgbench-data              Bound    pvc-0000e238-37ea-40d3-bedf-5bab59b1e019   10Gi       RWO            postgres-pgbench-sc   15m
pgbench-data-migrated     Bound    pvc-a2673fb4-e637-4aac-86b8-af60cad7a21f   10Gi       RWO            postgres-pgbench-sc   3m14s
pgbench-state             Bound    pvc-d7bbd9b1-18ae-4838-b2c9-169680b86628   1Gi        RWO            postgres-pgbench-sc   15m
pgbench-state1-migrated   Bound    pvc-82cfbf5e-322c-4785-b421-98324f8e0806   1Gi        RWO            postgres-pgbench-sc   3m14s
oksana@dev-onaumov ~/g/kubeup (master)>
```

File: staging/src/k8s.io/csi-translation-lib/plugins/porwotx_test.go
Comment (jsafrane) [inline staging/src/k8s.io/csi-translation-lib/plugins/porwotx_test.go:1]: There is a typo in the file name, `porwotx_test.go`
/hold

**Review by jsafrane — COMMENTED** [review]
/approve
from sig-storage perspective

File: staging/src/k8s.io/csi-translation-lib/plugins/portworx.go
Comment (msau42) [inline staging/src/k8s.io/csi-translation-lib/plugins/portworx.go:46]: You may also need to translate the intree fstype key to the csi fstype key: https://github.com/kubernetes/kubernetes/blob/6e56b766b4d1393bfa95013cb5fa7564be47c9c9/staging/src/k8s.io/csi-translation-lib/plugins/gce_pd.go#L87

**Comment by msau42** [comment]
Can you also update the release notes to indicate that this is alpha?

**Comment by trierra** [comment]
@msau42 done

File: staging/src/k8s.io/csi-translation-lib/plugins/portworx.go
Comment (trierra) [reply staging/src/k8s.io/csi-translation-lib/plugins/portworx.go:46]: In Portworx The actual flag is called `fs` and is the same across in-tree storage class and csi storage class

Comment (msau42) [reply staging/src/k8s.io/csi-translation-lib/plugins/portworx.go:46]: Ok CSI has defined a first class fstype key that allows it to get passed down as a CSI parameter. You may want to consider updating the driver to support that, but this is fine for now since the driver doesn't support it.

**Comment by msau42** [comment]
/assign derekwaynecarr, huang-wei, sttts
for approvals

File: staging/src/k8s.io/csi-translation-lib/plugins/porwotx_test.go
Comment (jsafrane) [reply staging/src/k8s.io/csi-translation-lib/plugins/porwotx_test.go]: I am really sorry to be so annoying, but the file name is still wrong :-). `porworx_test.go`

Comment (trierra) [reply staging/src/k8s.io/csi-translation-lib/plugins/porwotx_test.go]: oh, I'm sorry. Fixed

**Comment by ahg-g** [comment]
/approve 

for the scheduler

**Comment by ahg-g** [comment]
/hold

@msau42 should the feature gate be added to https://github.com/kubernetes/kubernetes/blob/ed3d9f7b48290d26ea617b2af40188e397987ac6/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1018 ?

same thing for https://github.com/kubernetes/kubernetes/pull/95361 ?

**Comment by msau42** [comment]
> should the feature gate be added to
> 
> https://github.com/kubernetes/kubernetes/blob/ed3d9f7b48290d26ea617b2af40188e397987ac6/pkg/scheduler/framework/plugins/volumebinding/binder.go#L1018
> 
> ?
> same thing for #95361 ?

Yes that's a good catch. The logic is there to translate the intree topology labels to the CSI topology labels. It's technically not necessary if a driver doesn't support topology but it' simpler if we keep the logic consistent with the other migrated plugins.

**Comment by sttts** [comment]
/approve

for the kcm changes.

**Comment by trierra** [comment]
@msau42 @ahg-g do you want me to add changes to the `kubernetes/pkg/scheduler/framework/plugins/volumebinding/binder.go` now or it can wait until beta?

**Comment by yujuhong** [comment]
/approve

for the cmd/kubelet change

**Pull request merged**
